src/server.cpp

```diff
@@ -908,6 +915,54 @@ void CServer::OnCLDisconnection ( CHostAddress InetAddr )
     }
 }
 
+void CServer::OnAboutToQuit()
+{
+    if ( IsRunning() )
+    {
+        // Clean up
+        CleanShutdown();
+    }
+
+    // if server was registered at the central server, unregister on shutdown
+    if ( GetServerListEnabled() )
+    {
+        UnregisterSlaveServer();
+    }
+}
+
+void CServer::OnShutdown ( int )
+{
+    // This should trigger OnAboutToQuit
+    QCoreApplication::instance()->exit();
+}
+
+
+void CServer::CleanShutdown() {
+    Mutex.lock();
+    {
+        bool oldDAC = bDisconnectAllClients;
+
+        // This is to prevent new connections
+        bDisconnectAllClients = true;
+
+        // This is to disconnect all connected channels
+        for ( int i = 0; i < iMaxNumChannels; i++ )
```

Comment:
This is critical because of two reasons:
I sometimes have to restart the Central Server even if musicians are connected to the server. Since the server restarts very quickly, the musicians only have a short period of time when they do not hear each other but after that they can continue their jam as if nothing happend. If you initiate a disconnect, all musicians would have to re-connect. I prefer the current behavior.
Second issue: The disconnect message was introduce by me just to disconnect clients which are connected over a long period of time to my server eating up my bandwidth. With the disconnect functionality I could restart the server with the --disconnectall and then I could restart the normal server operation and the client connection was gone. But the way the disconnect message from the server is treated in the client is not good right now. It was only intended for an "emergency case" and not for normal operation.

Reply:
Mmm, yes.  I see the conflict.

I don't actually need the client disconnect here.  (It didn't help the jam recorder, anyway.)  I was thinking it was "tidier" but it doesn't really let the client know what's going on anyway.  If the server has gone, the client will find out the same way whether or not disconnect is used.